session/session.go

```diff
@@ -1052,13 +1065,111 @@ func (s *session) SetGlobalSysVar(name, value string) error {
 		return err
 	}
 	name = strings.ToLower(name)
+	// update mysql.tidb if required.
+	if err = s.SetTiKVGlobalSysVar(name, sVal); err != nil {
+		return err
+	}
 	variable.CheckDeprecationSetSystemVar(s.sessionVars, name)
 	sql := fmt.Sprintf(`REPLACE %s.%s VALUES ('%s', '%s');`,
 		mysql.SystemDB, mysql.GlobalVariablesTable, name, sVal)
 	_, _, err = s.ExecRestrictedSQL(sql)
 	return err
 }
 
+// SetTiKVGlobalSysVar handles tikv_* sysvars which need to update mysql.tidb
+// for backwards compatibility. Validation has already been performed.
+func (s *session) SetTiKVGlobalSysVar(name, val string) error {
+	switch name {
+	case variable.TiKVGCConcurrency:
+		autoConcurrency := "false"
+		if val == "-1" {
+			autoConcurrency = "true"
+		}
+		sql := fmt.Sprintf(`INSERT INTO mysql.tidb (variable_name, variable_value, comment) VALUES ('%s', '%s', '%s')
+			ON DUPLICATE KEY UPDATE variable_value = '%s'`, "tikv_gc_auto_concurrency", autoConcurrency, gcVariableComments[name], autoConcurrency)
```

Comment:
Is there a method to safely escape a string? I looked for one, but could not find it.

In this case there is not an SQL injection risk because `autoConcurrency` and `gcVariableComments[name]` are not specified by users, but there are existing updates which I think could be made safer. Line `1098` does update `val` based on user input and needs to be escaped.

Reply:
It looks like `SET GLOBAL` is vulnerable, but since only `super` users can `SET GLOBAL` the risk is fairly low for now. But will need fixing. I'll file a separate bug:
```sql
mysql> set global sync_relay_log = "'";
ERROR 1064 (42000): You have an error in your SQL syntax; check the manual that corresponds to your TiDB version for the right syntax to use line 1 column 62 near "''');" 
```

Reply:
I have filed #22145 and added it to the list of fixed issues in this PR.